applications/StructuralMechanicsApplication/custom_processes/metrics_spr_error_process.cpp

```diff
@@ -0,0 +1,659 @@
+/***********************************************************************************/
+
+template<SizeType TDim>
+void SPRMetricProcess<TDim>::CalculateSuperconvergentStresses()
+{
+    FindNodalNeighboursProcess find_neighbours(mThisModelPart);
+    find_neighbours.Execute();
+
+    // Iteration over all nodes -- construction of patches
+    NodesArrayType& nodes_array = mThisModelPart.Nodes();
+    const int num_nodes = static_cast<int>(nodes_array.size());
+
+    #pragma omp parallel for
+    for(int i_node = 0; i_node < num_nodes; ++i_node) {
+        auto it_node = nodes_array.begin() + i_node;
+
+        KRATOS_DEBUG_ERROR_IF_NOT(it_node->Has(NEIGHBOUR_ELEMENTS)) << "SPRMetricProcess:: Search didn't work with elements" << std::endl;
+        const SizeType neighbour_size = it_node->GetValue(NEIGHBOUR_ELEMENTS).size();
+
+        Vector sigma_recovered(SigmaSize, 0.0);
+
+        if(neighbour_size > TDim) {
+            CalculatePatch(it_node, it_node, neighbour_size,sigma_recovered);
+            it_node->SetValue(RECOVERED_STRESS, sigma_recovered);
+
+            KRATOS_INFO_IF("SPRMetricProcess", mEchoLevel > 2) << "Recovered sigma: " << sigma_recovered << std::endl;
+        } else {
+            KRATOS_DEBUG_ERROR_IF_NOT(it_node->Has(NEIGHBOUR_NODES)) << "SPRMetricProcess:: Search didn't work with nodes" << std::endl;
+            auto& neigh_nodes = it_node->GetValue(NEIGHBOUR_NODES);
+            for(auto it_neighbour_nodes = neigh_nodes.begin(); it_neighbour_nodes != neigh_nodes.end(); it_neighbour_nodes++) {
+
+                Vector sigma_recovered_i(SigmaSize,0);
+
+                IndexType count_i = 0;
+                for(IndexType i_node_loop = 0; i_node_loop < num_nodes; ++i_node_loop) { // FIXME: Avoid this double loop, extreamily expensive
+                    auto it_node_loop = nodes_array.begin() + i_node_loop;
+                    const SizeType size_elem_neigh = it_node_loop->GetValue(NEIGHBOUR_ELEMENTS).size();
+                    if (it_node_loop->Id() == it_neighbour_nodes->Id() && size_elem_neigh > TDim){
+                        CalculatePatch(it_node, it_node_loop, neighbour_size, sigma_recovered_i);
+                        ++count_i;
+                    }
+                }
+
+                // Average solution from different patches
+                if(count_i != 0)
+                    sigma_recovered = sigma_recovered*(count_i-1)/count_i + sigma_recovered_i/count_i;
+            }
+
+            it_node->SetValue(RECOVERED_STRESS,sigma_recovered);
+            KRATOS_INFO_IF("SPRMetricProcess", mEchoLevel > 2) << "Recovered sigma: " << sigma_recovered << std::endl;
+        }
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<SizeType TDim>
+void SPRMetricProcess<TDim>::CalculateErrorEstimationAndElementSize(
```

Comment:
Long function. To shorten split the two loops into `CalculateErrorEstimation` and `CalculateElementSize` and call these here.

Reply:
+1 for @msandre

Reply:
actually... why do we need this function?

Reply:
To compute the metric